spec/p2p/reactor/reactor.md

```diff
@@ -0,0 +1,228 @@
+The typical behavior is to start routines that, given some conditions or events,
+send messages to the added peer, using the provided `Peer` instance.
+The companion [API for Reactors](./p2p-api.md#peer-api) documents the methods
+provided by `Peer` instances, available from when they are added to the reactors.
+
+## Stop Peer
+
+The p2p layer informs all registered reactors when it disconnects from a `Peer`,
+using the `RemovePeer(Peer, reason)` method:
+
+```abnf
+stop-peer       = [peer-error] remove-peer
+```
+
+This method is invoked after the p2p layer has stopped peer's send and receive routines.
+Depending of the `reason` for which the peer was stopped, different log
+messages can be produced.
+After removing a peer from all reactors, the `Peer` instance is also removed from
+the set of connected peers.
+This enables the same peer to reconnect and `InitPeer(Peer)` to be invoked for
+the new connection.
+
+From the removal of a `Peer` , the reactor should not receive any further message
+from the peer and must not try sending messages to the removed peer.
+This usually means stopping the routines that were started by the companion
+`Add(Peer)` method.
+
+## Receive messages
+
+The main duty of a reactor is to handle incoming messages on the channels it
+has registered with the p2p layer.
+
+The _pre-condition_ for receiving a message from a `Peer` is that the p2p layer
+has previously invoked `InitPeer(Peer)`.
+This means that the reactor must be able to receive a message from a `Peer`
+_before_ `AddPeer(Peer)` is invoked.
+This happens because the peer's send and receive routines are started before,
+and should be already running when the p2p layer adds the peer to every
+registered reactor.
+
+```abnf
+start-peer      = [*receive] (connected-peer / start-error)
+connected-peer  = add-peer *receive
+```
+
+The most common scenario, however, is to start receiving messages from a peer
+after `AddPeer(Peer)` is invoked.
+An arbitrary number of messages can be received, until the peer is stopped and
+`RemovePeer(Peer)` is invoked.
+
+When a message is received from a connected peer on any of the channels
+registered by the reactor, the p2p layer will deliver the message to the
+reactor via the `Receive(Envelope)` method.
+The message is packed into an `Envelope` that contains:
+
+- `ChannelID`: the channel the message belongs to
+- `Src`: the source `Peer` handler, from which the message was received
+- `Message`: the actual message's payload, unmarshalled using protocol buffers
+
+Two important observations regarding the implementation of the `Receive` method:
+
+1. Concurrency: the implementation should consider concurrent invocations of
+   the `Receive` method carrying messages from different peers, as the
+   interaction with different peers is independent and messages can be received in parallel.
+1. Non-blocking: the implementation of the `Receive` method is expected not to block,
```

Comment:
I don't think "non-blocking" is the right term here, as the function is blocking, be it quick or not.
Not sure what a better term would be. "Does not block indefinitely", maybe?